src/com/palmergames/bukkit/towny/listeners/TownyEntityListener.java

```diff
@@ -313,51 +315,55 @@ public void onPotionSplashEvent(PotionSplashEvent event) {
 		if (!TownyAPI.getInstance().isTownyWorld(event.getEntity().getWorld()))
 			return;
 		
-		List<LivingEntity> affectedEntities = (List<LivingEntity>) event.getAffectedEntities();
 		ThrownPotion potion = event.getPotion();
 		Entity attacker = null;
 
-		List<PotionEffect> effects = (List<PotionEffect>) potion.getEffects();
 		boolean detrimental = false;
 
 		/*
 		 * List of potion effects blocked from PvP.
 		 */
-		List<String> prots = TownySettings.getPotionTypes();
+		List<String> detrimentalPotions = TownySettings.getPotionTypes();
 		
-		
-		for (PotionEffect effect : effects) {
+		for (PotionEffect effect : potion.getEffects()) {
 
 			/*
 			 * Check to see if any of the potion effects are protected.
 			 */
 			
-			if (prots.contains(effect.getType().getName())) {
+			if (detrimentalPotions.contains(effect.getType().getName())) {
 				detrimental = true;
+				break;
 			}
-
 		}
+		
+		if (!detrimental)
+			return;
 
 		Object source = potion.getShooter();
 		Block dispenser = null;
 
 		if (source instanceof BlockProjectileSource blockProjectileSource) {
 			dispenser = blockProjectileSource.getBlock();
+		} else if (source instanceof Entity entity) {
+			attacker = entity;
 		} else {
-			attacker = (Entity) source;
+			// The source is not a block (dispenser) and not an entity somehow.
+			// Probably caused by a faulty "custom entities" plugin.
+			return;

```

Comment:
I think instead of returning we should treat it similar to how the CombatUtil.preventDispenserDamage() is done, but instead of passing a Block we pass a Potion or Entity. Ultimately it would test the defender's location for PVP activity using CombatUtil#isPVP(TownBlock) or TownyAPI#isPVP(Location). That would mean that the pvp events are at least thrown.

Or we can do a whole custom method like preventDispenserDamage with its own TownyDamageEvent event.

Since its a detrimental potion we ought to be doing something about it.